<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Max Conway" />

<meta name="date" content="2020-09-02" />

<title>Multi-Objective Optimization case study</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Multi-Objective Optimization case study</h1>
<h4 class="author">Max Conway</h4>
<h4 class="date">2020-09-02</h4>



<div id="multi-objective-optimization-tutorial" class="section level1">
<h1>Multi Objective Optimization tutorial</h1>
<p>This tutorial is an in depth example of the use of this package in the context of an evolutionary optimization approach. It also shows the use of <code>gene_eval</code> to calculate a reaction speed from gene expressions.</p>
<p>First, we need to load the appropriate libraries. You may also need to load your optimization library of choice.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(dplyr)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">library</span>(purrr)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">library</span>(stringr)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw">library</span>(fbar)</span></code></pre></div>
<p>This code block loads a model, then extracts the list of genes from the model. The model takes the form of a tabular list of reactions.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">data</span>(<span class="st">&quot;ecoli_core&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a>model &lt;-<span class="st"> </span>ecoli_core <span class="op">%&gt;%</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">geneAssociation =</span> geneAssociation <span class="op">%&gt;%</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="st">           </span><span class="kw">str_replace_all</span>(<span class="st">&#39;and&#39;</span>,<span class="st">&#39;&amp;&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="st">           </span><span class="kw">str_replace_all</span>(<span class="st">&#39;or&#39;</span>,<span class="st">&#39;|&#39;</span>))</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a>genes_in_model &lt;-<span class="st"> </span>model<span class="op">$</span>geneAssociation <span class="op">%&gt;%</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="st">  </span><span class="kw">str_split</span>(<span class="st">&#39;[()|&amp; ]+&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="st">  </span><span class="kw">flatten_chr</span>() <span class="op">%&gt;%</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="st">  </span><span class="kw">discard</span>(is.na) <span class="op">%&gt;%</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="st">  </span><span class="kw">discard</span>(<span class="op">~</span><span class="st"> </span><span class="kw">str_length</span>(.x)<span class="op">==</span><span class="dv">0</span>)</span></code></pre></div>
<p>The evaluation function is where the actual metabolic simulations are performed. This has four main stages:</p>
<ul>
<li>The gene-reaction associations (<code>geneAssociation</code>) are evaluated in the context of which genes are present in this iteration (<code>genome</code>).</li>
<li>We conduct a round of FBA, optimizing for maximum biomass.</li>
<li>Having found the maximum biomass production, we fix the biomass at this value (+/-1%).</li>
<li>With the biomass value fixed, we the optimize to maximize the synthetic objective.</li>
</ul>
<p>The technique of fixing the biomass followed by maximizing the synthetic objective is important because there could still be slack in the model after the first optimization stage, and we wish to have a reliable synthetic objective estimate.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>evaluation_function &lt;-<span class="st"> </span><span class="cf">function</span>(genome){</span>
<span id="cb3-2"><a href="#cb3-2"></a>  </span>
<span id="cb3-3"><a href="#cb3-3"></a>  res &lt;-<span class="st"> </span>model <span class="op">%&gt;%</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">activation =</span> <span class="kw">gene_eval</span>(geneAssociation, <span class="kw">names</span>(genome), genome),</span>
<span id="cb3-5"><a href="#cb3-5"></a>           <span class="dt">activation =</span> <span class="kw">coalesce</span>(activation, <span class="dv">1</span>),</span>
<span id="cb3-6"><a href="#cb3-6"></a>           <span class="dt">uppbnd =</span> uppbnd<span class="op">*</span>activation,</span>
<span id="cb3-7"><a href="#cb3-7"></a>           <span class="dt">lowbnd =</span> lowbnd<span class="op">*</span>activation) <span class="op">%&gt;%</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="st">    </span><span class="kw">find_fluxes_df</span>(<span class="dt">do_minimization =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">lowbnd =</span> <span class="kw">ifelse</span>(abbreviation<span class="op">==</span><span class="st">&#39;BIOMASS_Ecoli_core_w_GAM&#39;</span>, flux<span class="op">*</span><span class="fl">0.99</span>, lowbnd),</span>
<span id="cb3-10"><a href="#cb3-10"></a>           <span class="dt">uppbnd =</span> <span class="kw">ifelse</span>(abbreviation<span class="op">==</span><span class="st">&#39;BIOMASS_Ecoli_core_w_GAM&#39;</span>, flux<span class="op">*</span><span class="fl">1.01</span>, uppbnd),</span>
<span id="cb3-11"><a href="#cb3-11"></a>           <span class="dt">obj_coef =</span> <span class="dv">1</span><span class="op">*</span>(abbreviation<span class="op">==</span><span class="st">&#39;EX_ac_e&#39;</span>)) <span class="op">%&gt;%</span></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="st">    </span><span class="kw">find_fluxes_df</span>(<span class="dt">do_minimization =</span> <span class="ot">FALSE</span>)</span>
<span id="cb3-13"><a href="#cb3-13"></a>  </span>
<span id="cb3-14"><a href="#cb3-14"></a>  <span class="kw">return</span>(<span class="kw">list</span>(<span class="dt">bm =</span> <span class="kw">filter</span>(res, abbreviation<span class="op">==</span><span class="st">&#39;BIOMASS_Ecoli_core_w_GAM&#39;</span>)<span class="op">$</span>flux, </span>
<span id="cb3-15"><a href="#cb3-15"></a>              <span class="dt">synth =</span> <span class="kw">filter</span>(res, abbreviation<span class="op">==</span><span class="st">&#39;EX_ac_e&#39;</span>)<span class="op">$</span>flux))</span>
<span id="cb3-16"><a href="#cb3-16"></a>}</span></code></pre></div>
<p>Non-domination sorting is the first stage of the selection procedure in NSGA-II. The code might be quite opaque, but the idea is as follows:</p>
<ul>
<li>We perform an inner_join in order to compare every point against every other point.</li>
<li>For each point (<code>id.x</code>), we see if there exists any second point (<code>id.y</code>) that has a higher value than it in all objectives. Where such a second point exists, we term the original point ‘dominated’.</li>
<li>We find the set of points which have no dominating point, and term this the first non-dominated front.</li>
<li>We repeat this procedure, but ignoring points in the first non-dominated front, to find the second on-dominated front, and so on.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>non_dom_sort &lt;-<span class="st"> </span><span class="cf">function</span>(input){</span>
<span id="cb4-2"><a href="#cb4-2"></a>  input_long &lt;-<span class="st"> </span>input <span class="op">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="st">    </span><span class="kw">gather</span>(property, value, <span class="op">-</span>id) <span class="op">%&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">front=</span><span class="ot">NA</span>)</span>
<span id="cb4-5"><a href="#cb4-5"></a>  </span>
<span id="cb4-6"><a href="#cb4-6"></a>  currentfront &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb4-7"><a href="#cb4-7"></a>  </span>
<span id="cb4-8"><a href="#cb4-8"></a>  <span class="cf">while</span>(<span class="kw">any</span>(<span class="kw">is.na</span>(input_long<span class="op">$</span>front))){</span>
<span id="cb4-9"><a href="#cb4-9"></a>    </span>
<span id="cb4-10"><a href="#cb4-10"></a>    input_long &lt;-<span class="st"> </span>input_long <span class="op">%&gt;%</span></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="st">      </span><span class="kw">inner_join</span>(.,., <span class="dt">by=</span><span class="st">&#39;property&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="st">      </span><span class="kw">group_by</span>(id.x,id.y) <span class="op">%&gt;%</span></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="st">      </span><span class="kw">mutate</span>(<span class="dt">dominance =</span> <span class="kw">ifelse</span>(<span class="kw">all</span>(value.x<span class="op">&gt;=</span>value.y), </span>
<span id="cb4-14"><a href="#cb4-14"></a>                                <span class="st">&#39;xdomy&#39;</span>, </span>
<span id="cb4-15"><a href="#cb4-15"></a>                                <span class="kw">ifelse</span>(<span class="kw">all</span>(value.y<span class="op">&gt;=</span>value.x), </span>
<span id="cb4-16"><a href="#cb4-16"></a>                                       <span class="st">&#39;ydomx&#39;</span>, </span>
<span id="cb4-17"><a href="#cb4-17"></a>                                       <span class="st">&#39;nondom&#39;</span></span>
<span id="cb4-18"><a href="#cb4-18"></a>                                       )</span>
<span id="cb4-19"><a href="#cb4-19"></a>                                )</span>
<span id="cb4-20"><a href="#cb4-20"></a>      ) <span class="op">%&gt;%</span></span>
<span id="cb4-21"><a href="#cb4-21"></a><span class="st">      </span><span class="kw">group_by</span>(id.x) <span class="op">%&gt;%</span></span>
<span id="cb4-22"><a href="#cb4-22"></a><span class="st">      </span><span class="kw">mutate</span>(<span class="dt">front =</span> <span class="kw">ifelse</span>(<span class="kw">all</span>(dominance[<span class="kw">is.na</span>(front.y)] <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;xdomy&#39;</span>, <span class="st">&#39;nondom&#39;</span>)), </span>
<span id="cb4-23"><a href="#cb4-23"></a>                            <span class="kw">pmin</span>(currentfront, front.x, <span class="dt">na.rm=</span><span class="ot">TRUE</span>), </span>
<span id="cb4-24"><a href="#cb4-24"></a>                            <span class="ot">NA</span></span>
<span id="cb4-25"><a href="#cb4-25"></a>                            )</span>
<span id="cb4-26"><a href="#cb4-26"></a>      ) <span class="op">%&gt;%</span></span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="st">      </span><span class="kw">group_by</span>(<span class="dt">id =</span> id.x, <span class="dt">property =</span> property, front, <span class="dt">value =</span> value.x) <span class="op">%&gt;%</span></span>
<span id="cb4-28"><a href="#cb4-28"></a><span class="st">      </span><span class="kw">summarise</span>()</span>
<span id="cb4-29"><a href="#cb4-29"></a>    </span>
<span id="cb4-30"><a href="#cb4-30"></a>    currentfront &lt;-<span class="st"> </span>currentfront <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb4-31"><a href="#cb4-31"></a>  }</span>
<span id="cb4-32"><a href="#cb4-32"></a>  </span>
<span id="cb4-33"><a href="#cb4-33"></a>  <span class="kw">return</span>(</span>
<span id="cb4-34"><a href="#cb4-34"></a>    input_long <span class="op">%&gt;%</span></span>
<span id="cb4-35"><a href="#cb4-35"></a><span class="st">      </span><span class="kw">spread</span>(property, value)</span>
<span id="cb4-36"><a href="#cb4-36"></a>  )</span>
<span id="cb4-37"><a href="#cb4-37"></a>  </span>
<span id="cb4-38"><a href="#cb4-38"></a>}</span></code></pre></div>
<p>The second part of the NSGA-II evaluation procedure is finding the crowding distance. This is used to break ties between points in the same non-dominated front. In for each front, for each dimension, this function sorts the points into order along the dimension, and finds the normalized distance between the proceeding point and succeeding point. These values are summed up across each dimension to find the value for the point.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>crowding_distance &lt;-<span class="st"> </span><span class="cf">function</span>(input){</span>
<span id="cb5-2"><a href="#cb5-2"></a>  <span class="kw">return</span>(</span>
<span id="cb5-3"><a href="#cb5-3"></a>    input <span class="op">%&gt;%</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="st">      </span><span class="kw">gather</span>(property, value, <span class="op">-</span>id, <span class="op">-</span>front) <span class="op">%&gt;%</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="st">      </span><span class="kw">group_by</span>(front, property) <span class="op">%&gt;%</span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="st">      </span><span class="kw">arrange</span>(value) <span class="op">%&gt;%</span></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="st">      </span><span class="kw">mutate</span>(<span class="dt">crowding =</span> (<span class="kw">lead</span>(value)<span class="op">-</span><span class="kw">lag</span>(value))<span class="op">/</span>(<span class="kw">max</span>(value)<span class="op">-</span><span class="kw">min</span>(value)),</span>
<span id="cb5-8"><a href="#cb5-8"></a>             <span class="dt">crowding =</span> <span class="kw">ifelse</span>(<span class="kw">is.na</span>(crowding),<span class="ot">Inf</span>, crowding)) <span class="op">%&gt;%</span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="st">      </span><span class="kw">group_by</span>(id) <span class="op">%&gt;%</span></span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="st">      </span><span class="kw">mutate</span>(<span class="dt">crowding =</span> <span class="kw">sum</span>(crowding)) <span class="op">%&gt;%</span></span>
<span id="cb5-11"><a href="#cb5-11"></a><span class="st">      </span><span class="kw">spread</span>(property, value)</span>
<span id="cb5-12"><a href="#cb5-12"></a>  )</span>
<span id="cb5-13"><a href="#cb5-13"></a>}</span></code></pre></div>
<p>This is the genetic loop of the algorithm. It is explained by code comments, but follows a normal pattern of evaluating, sorting, selecting from and mutating the population.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>start_genome &lt;-<span class="st"> </span><span class="kw">set_names</span>(<span class="kw">rep_along</span>(genes_in_model, <span class="ot">TRUE</span>), genes_in_model)</span>
<span id="cb6-2"><a href="#cb6-2"></a>pop &lt;-<span class="st"> </span><span class="kw">list</span>(start_genome)</span>
<span id="cb6-3"><a href="#cb6-3"></a></span>
<span id="cb6-4"><a href="#cb6-4"></a>popsize =<span class="st"> </span><span class="dv">50</span></span>
<span id="cb6-5"><a href="#cb6-5"></a></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">50</span>){</span>
<span id="cb6-7"><a href="#cb6-7"></a>  results &lt;-<span class="st"> </span><span class="kw">map_df</span>(pop, evaluation_function) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Evaluate all the genomes</span></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">bm=</span><span class="kw">signif</span>(bm), <span class="dt">synth=</span><span class="kw">signif</span>(synth)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Round results</span></span>
<span id="cb6-9"><a href="#cb6-9"></a><span class="st">    </span><span class="kw">unique</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Throw away duplicates</span></span>
<span id="cb6-10"><a href="#cb6-10"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">id =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">n</span>()) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># label the results</span></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="st">    </span><span class="kw">sample_frac</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Shuffle</span></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="st">    </span><span class="kw">non_dom_sort</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Find the non-dominated fronts</span></span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="st">    </span>crowding_distance <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Find the crowding distances</span></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="st">    </span><span class="kw">arrange</span>(front, <span class="kw">desc</span>(crowding)) <span class="co"># Sort by front, breaking ties by crowding distance</span></span>
<span id="cb6-15"><a href="#cb6-15"></a>  </span>
<span id="cb6-16"><a href="#cb6-16"></a>  selected &lt;-<span class="st"> </span>results <span class="op">%&gt;%</span></span>
<span id="cb6-17"><a href="#cb6-17"></a><span class="st">    </span><span class="kw">filter</span>(<span class="kw">row_number</span>() <span class="op">&lt;=</span><span class="st"> </span>popsize<span class="op">/</span><span class="dv">2</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Keep the best half of the population</span></span>
<span id="cb6-18"><a href="#cb6-18"></a><span class="st">    </span><span class="kw">getElement</span>(<span class="st">&#39;id&#39;</span>)</span>
<span id="cb6-19"><a href="#cb6-19"></a>  </span>
<span id="cb6-20"><a href="#cb6-20"></a>  kept_pop &lt;-<span class="st"> </span>pop[selected]</span>
<span id="cb6-21"><a href="#cb6-21"></a>  altered_pop &lt;-<span class="st"> </span>kept_pop <span class="op">%&gt;%</span></span>
<span id="cb6-22"><a href="#cb6-22"></a><span class="st">    </span><span class="kw">sample</span>(popsize<span class="op">-</span><span class="kw">length</span>(selected), <span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Select a random portion of the population as parents</span></span>
<span id="cb6-23"><a href="#cb6-23"></a><span class="st">    </span><span class="kw">map</span>(<span class="cf">function</span>(genome){<span class="kw">xor</span>(genome, <span class="kw">runif</span>(<span class="kw">length</span>(genome))<span class="op">&gt;</span><span class="fl">0.98</span>)}) <span class="co"># Mutate a small number of genes from the parent population.</span></span>
<span id="cb6-24"><a href="#cb6-24"></a>  </span>
<span id="cb6-25"><a href="#cb6-25"></a>  pop &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">c</span>(kept_pop, altered_pop)) <span class="co"># Combine the ofspring and parent populations</span></span>
<span id="cb6-26"><a href="#cb6-26"></a>}</span></code></pre></div>
<p>Now that we have results, the set of all non-dominated points, known as the Pareto Front. This describes the tradeoff between biomass production and our synthetic objective.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a>results <span class="op">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="st">  </span><span class="kw">filter</span>(front<span class="op">==</span><span class="dv">1</span>) <span class="op">%&gt;%</span></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>bm, <span class="dt">y=</span>synth, <span class="dt">colour=</span>front)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="st">  </span><span class="kw">geom_step</span>(<span class="dt">direction=</span><span class="st">&#39;vh&#39;</span>) <span class="op">+</span></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="st">  </span><span class="kw">scale_x_log10</span>() <span class="op">+</span></span>
<span id="cb7-9"><a href="#cb7-9"></a><span class="st">  </span><span class="kw">scale_y_log10</span>() <span class="op">+</span></span>
<span id="cb7-10"><a href="#cb7-10"></a><span class="st">  </span><span class="kw">theme_bw</span>()</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
